File: sdks/rust/src/gate.rs
//! Paywall gate ergonomics (§2.4) — outcomes and trackers delegate decisions to core.

#![allow(clippy::missing_docs_in_private_items)]

use crate::client::Client;
use serde_json::Value;
use solvapay_core::{PaywallGate, SdkError};

/// Options for [`Client::gate`].
#[derive(Debug, Clone)]
pub struct GateOpts {
    /// Product reference (`productRef` on limit checks).
    pub product: String,
    /// Meter / usage type (default `"requests"`).
    pub usage_type: String,
}

impl GateOpts {
    /// Builds gate options for a product with the default usage type.
    ///
    /// # Arguments
    ///
    /// * `product` - Product reference.
    ///
    /// # Returns
    ///
    /// Options with `usage_type` set to `"requests"`.
    pub fn for_product(product: impl Into<String>) -> Self {
        Self {
            product: product.into(),
            usage_type: "requests".to_owned(),
        }
    }
}

/// Result of a paywall gate check.
///
/// `Allow` holds the live client plus last limits/customer snapshot; boxing it
/// would change the public match shape for no integrator benefit.
#[allow(clippy::large_enum_variant)]
pub enum GateOutcome {
    /// Request is gated; render 402 from structured gate content.
    Paywall(PaywallGate),
    /// Request may proceed; call [`Allow::track_success`] / [`Allow::track_fail`] after work.
    Allow(Allow),
}

impl From<solvapay_core::CustomerSnapshot> for CustomerSnapshot {
    fn from(customer: solvapay_core::CustomerSnapshot) -> Self {
        Self {
            customer_ref: customer.customer_ref,
            balance: serde_json::json!(customer.balance),
            remaining: customer.remaining,
            within_limits: serde_json::json!(customer.within_limits),
            plan: customer.plan,
            throttled: customer.throttled,
            overage: customer.overage,
        }
    }
}

/// Merchant-facing customer projection from the last limits check.
#[derive(Debug, Clone, PartialEq)]
pub struct CustomerSnapshot {
    /// Backend customer reference used for usage tracking.
    pub customer_ref: String,
    /// `creditBalance` from limits, or `0` when absent.
    pub balance: Value,
    /// Remaining allowance from limits (may be JSON `null`).
    pub remaining: Value,
    /// `withinLimits` from limits, or `true` when absent.
    pub within_limits: Value,
    /// Plan field from limits (may be JSON `null`).
    pub plan: Value,
    /// `throttled` from limits, or `false` when absent.
    pub throttled: bool,
    /// `overage` from limits, or `false` when absent.
    pub overage: bool,
}

/// Allow arm returned from [`Client::gate`]; usage tracking delegates to the typed client.
pub struct Allow {
    pub(crate) client: Client,
    pub(crate) backend_ref: String,
    pub(crate) product: String,
    pub(crate) meter_name: String,
    pub(crate) limits: Value,
    pub(crate) customer: CustomerSnapshot,
    pub(crate) consequence: Option<solvapay_core::AllowConsequence>,
    pub(crate) driver_state: Value,
}

/// Classifies a [`Allow::track_fail`] cause for `gate_next` (`isPaywallError`).
pub trait TrackFailCause {
    /// Human-readable fail message forwarded to `gate_next`.
    fn track_fail_message(&self) -> String;

    /// Whether this cause is a paywall gate error (usage must be skipped).
    fn is_paywall_error(&self) -> bool {
        false
    }
}

impl TrackFailCause for str {
    fn track_fail_message(&self) -> String {
        self.to_owned()
    }
}

impl TrackFailCause for String {
    fn track_fail_message(&self) -> String {
        self.clone()
    }
}

impl TrackFailCause for SdkError {
    fn track_fail_message(&self) -> String {
        match self {
            SdkError::Api { message, .. }
            | SdkError::Paywall { message, .. }
            | SdkError::Transport { message, .. } => message.clone(),
            other => format!("{other:?}"),
        }
    }

    fn is_paywall_error(&self) -> bool {
        matches!(self, SdkError::Paywall { .. })
    }
}

impl<T: TrackFailCause + ?Sized> TrackFailCause for &T {
    fn track_fail_message(&self) -> String {
        (**self).track_fail_message()
    }

    fn is_paywall_error(&self) -> bool {
        (**self).is_paywall_error()
    }
}

/// Options for usage tracking after an allowed request.
#[derive(Debug, Clone, Default)]
pub struct TrackOpts {
    /// Optional duration in milliseconds.
    pub duration: Option<f64>,
    /// Optional metadata map (merged on track-fail with an `error` key).
    pub metadata: Option<serde_json::Map<String, serde_json::Value>>,
}

impl Allow {
    /// Customer snapshot used by payable MCP `ResponseContext`.
    pub fn customer(&self) -> CustomerSnapshot {
        self.customer.clone()
    }

    /// Degraded allow reason from the core driver. `None` on a plain allow.
    pub fn consequence(&self) -> Option<solvapay_core::AllowConsequence> {
        self.consequence
    }

    /// Backend customer reference used for usage tracking.
    pub fn backend_ref(&self) -> &str {
        &self.backend_ref
    }

    /// Product reference this allow decision was issued for.
    pub fn product(&self) -> &str {
        &self.product
    }

    /// Meter name used when recording usage for this allow.
    pub fn meter_name(&self) -> &str {
        &self.meter_name
    }

    /// Limits payload from the last successful check.
    pub fn limits(&self) -> &Value {
        &self.limits
    }

    /// Copy the core driver snapshot onto the host `CustomerSnapshot`.
    pub fn from_core_customer(customer: solvapay_core::CustomerSnapshot) -> CustomerSnapshot {
        CustomerSnapshot::from(customer)
    }

    /// Records a successful usage event (`trackUsage`).
    pub async fn track_success(&self, opts: TrackOpts) -> Result<(), solvapay_core::SdkError> {
        self.client
            .emit_handler_usage(
                &self.driver_state,
                serde_json::json!({
                    "kind": "handlerSucceeded",
                    "durationMs": opts.duration.unwrap_or(0.0),
                    "nowMs": crate::client::now_ms(),
                    "randomUnit": self.client.random_unit(),
                }),
            )
            .await
    }

    /// Records a failed usage event (same API call with error metadata).
    pub async fn track_fail(
        &self,
        error: impl TrackFailCause,
        opts: TrackOpts,
    ) -> Result<(), solvapay_core::SdkError> {
        self.client
            .emit_handler_usage(
                &self.driver_state,
                serde_json::json!({
                    "kind": "handlerFailed",
                    "durationMs": opts.duration.unwrap_or(0.0),
                    "nowMs": crate::client::now_ms(),
                    "randomUnit": self.client.random_unit(),
                    "errorMessage": error.track_fail_message(),
                    "isPaywallError": error.is_paywall_error(),
                }),
            )
            .await
    }
}

/// Product-scoped gate helper from [`Client::payable`].
#[derive(Clone)]
pub struct Payable {
    pub(crate) client: Client,
    pub(crate) product: String,
    pub(crate) usage_type: String,
}

impl Payable {
    /// Runs [`Client::gate`] with this payable's product and usage type.
    pub async fn gate(&self, customer_ref: &str) -> Result<GateOutcome, solvapay_core::SdkError> {
        self.client
            .gate(
                customer_ref,
                GateOpts {
                    product: self.product.clone(),
                    usage_type: self.usage_type.clone(),
                },
            )
            .await
    }
}

File: sdks/go/gate.go
package solvapay

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/solvapay/solvapay-go/internal/nativecall"
)

const defaultLimitsCacheTTL = time.Duration(DefaultLimitsCacheTTLMs) * time.Millisecond

// GateOpts configures Client.Gate.
type GateOpts struct {
	Product   string
	UsageType string
}

// GateOutcome is the result of a paywall gate check.
type GateOutcome interface{ isGateOutcome() }

// Paywall is the gated arm; Gate is the layer-2 structured content, verbatim.
type Paywall struct {
	Gate json.RawMessage
}

func (*Paywall) isGateOutcome() {}

// Allow is the proceed arm; call TrackSuccess / TrackFail after the handler.
type Allow struct {
	client      *Client
	backendRef  string
	product     string
	meterName   string
	limits      map[string]any
	customer    CustomerSnapshot
	driverState any
}

func (*Allow) isGateOutcome() {}

// TrackOpts configures usage tracking after an allowed request.
type TrackOpts struct {
	Duration *float64
	Metadata map[string]any
}

// CustomerSnapshot is the merchant-facing customer projection from the last limits check.
type CustomerSnapshot struct {
	Ref          string
	Balance      any
	Remaining    any
	WithinLimits any
	Plan         any
}

// Customer returns the snapshot used by payable MCP ResponseContext.
func (a *Allow) Customer() CustomerSnapshot {
	if a == nil {
		return CustomerSnapshot{}
	}
	return a.customer
}

func customerSnapshotFromAction(action map[string]any, backendRef string) (CustomerSnapshot, error) {
	raw, ok := action["customer"].(map[string]any)
	if !ok {
		return CustomerSnapshot{}, &Error{Code: "internal_error", Message: "gate_next missing customer snapshot"}
	}
	ref, _ := raw["ref"].(string)
	if ref == "" {
		ref = backendRef
	}
	return CustomerSnapshot{Ref: ref, Balance: raw["balance"], Remaining: raw["remaining"], WithinLimits: raw["withinLimits"], Plan: raw["plan"]}, nil
}

// TrackSuccess records a successful usage event.
func (a *Allow) TrackSuccess(ctx context.Context, opts TrackOpts) error {
	duration := 0.0
	if opts.Duration != nil {
		duration = *opts.Duration
	}
	return a.client.emitHandlerUsage(ctx, a.driverState, map[string]any{
		"kind":       "handlerSucceeded",
		"durationMs": duration,
		"nowMs":      time.Now().UnixMilli(),
		"randomUnit": mustRandomUnit(),
	})
}

// TrackFail records a failed usage event.
func (a *Allow) TrackFail(ctx context.Context, cause error, opts TrackOpts) error {
	duration := 0.0
	if opts.Duration != nil {
		duration = *opts.Duration
	}
	msg := ""
	if cause != nil {
		msg = cause.Error()
	}
	return a.client.emitHandlerUsage(ctx, a.driverState, map[string]any{
		"kind":           "handlerFailed",
		"durationMs":     duration,
		"nowMs":          time.Now().UnixMilli(),
		"randomUnit":     mustRandomUnit(),
		"errorMessage":   msg,
		"isPaywallError": IsPaywallError(cause),
	})
}

// Payable is a product-scoped gate helper.
type Payable struct {
	client    *Client
	product   string
	usageType string
}

// Payable returns a product-scoped gate helper.
func (c *Client) Payable(product, usageType string) *Payable {
	if usageType == "" {
		usageType = "requests"
	}
	return &Payable{client: c, product: product, usageType: usageType}
}

// Gate runs Client.Gate with this payable's product and usage type.
func (p *Payable) Gate(ctx context.Context, customerRef string) (GateOutcome, error) {
	return p.client.Gate(ctx, customerRef, GateOpts{Product: p.product, UsageType: p.usageType})
}

type gateState struct {
	mu sync.Mutex

	customerCache    map[string]customerCacheEntry
	customerInflight map[string]*customerInflight
	limitsCache      map[string]limitsCacheEntry
}

type customerCacheEntry struct {
	value       string
	timestampMs int64
}

type customerInflight struct {
	done chan struct{}
	ref  string
	err  error
}

type limitsCacheEntry struct {
	storedAt  time.Time
	remaining any
	limits    map[string]any
}

func newGateState() *gateState {
	return &gateState{
		customerCache:    map[string]customerCacheEntry{},
		customerInflight: map[string]*customerInflight{},
		limitsCache:      map[string]limitsCacheEntry{},
	}
}

// Gate classifies the customer, checks limits, and returns paywall or allow.
func (c *Client) Gate(ctx context.Context, customerRef string, opts GateOpts) (GateOutcome, error) {
	if c.gate == nil {
		c.gate = newGateState()
	}
	if opts.UsageType == "" {
		opts.UsageType = "requests"
	}
	if opts.Product == "" {
		return nil, &Error{Code: "invalid_config", Message: "product is required"}
	}
	startedMs := time.Now().UnixMilli()
	var state any
	event := map[string]any{
		"kind":             "start",
		"customerRef":      customerRef,
		"product":          opts.Product,
		"usageType":        opts.UsageType,
		"startedMs":        startedMs,
		"limitsCacheTTLMs": defaultLimitsCacheTTL.Milliseconds(),
	}
	var action map[string]any
	for {
		outJSON, err := callDecisionJSON(ctx, "sv_gate_next_binding", map[string]any{
			"state": state,
			"event": event,
		})
		if err != nil {
			return nil, err
		}
		var out struct {
			State  any            `json:"state"`
			Action map[string]any `json:"action"`
		}
		if err := json.Unmarshal(outJSON, &out); err != nil {
			return nil, fmt.Errorf("solvapay: gate_next: %w", err)
		}
		state = out.State
		action = out.Action
		kind, _ := action["kind"].(string)
		switch kind {
		case "ensureCustomer":
			ref, _ := action["customerRef"].(string)
			backendRef, err := c.ensureCustomer(ctx, ref)
			if err != nil {
				return nil, err
			}
			event = map[string]any{
				"kind":       "customerResolved",
				"backendRef": backendRef,
				"nowMs":      time.Now().UnixMilli(),
			}
		case "readLimitsCache":
			key, _ := action["key"].(string)
			now := time.Now().UnixMilli()
			c.gate.mu.Lock()
			cached, hit := c.gate.limitsCache[key]
			c.gate.mu.Unlock()
			if hit {
				event = map[string]any{
					"kind":        "limitsCacheEntry",
					"randomUnit":  mustRandomUnit(),
					"found":       true,
					"remaining":   cached.remaining,
					"limits":      cached.limits,
					"timestampMs": cached.storedAt.UnixMilli(),
					"nowMs":       now,
				}
			} else {
				event = map[string]any{
					"kind":       "limitsCacheEntry",
					"found":      false,
					"nowMs":      now,
					"randomUnit": mustRandomUnit(),
				}
			}
		case "checkLimits":
			if deleteKey, ok := action["cacheDeleteKey"].(string); ok && deleteKey != "" {
				c.gate.mu.Lock()
				delete(c.gate.limitsCache, deleteKey)
				c.gate.mu.Unlock()
			}
			raw, err := c.CheckLimits(ctx, map[string]any{
				"customerRef":            action["customerRef"],
				"productRef":             action["productRef"],
				"meterName":              action["meterName"],
				"includeCheckoutSession": asBool(action["includeCheckoutSession"]),
			})
			if err != nil {
				return nil, err
			}
			limits, ok := raw.(map[string]any)
			if !ok {
				return nil, fmt.Errorf("solvapay: checkLimits returned a non-object body")
			}
			event = map[string]any{
				"kind":       "limitsResult",
				"limits":     limits,
				"nowMs":      time.Now().UnixMilli(),
				"randomUnit": mustRandomUnit(),
			}
		case "allow":
			if err := c.applyGateCache(action["cache"]); err != nil {
				return nil, err
			}
			backendRef, _ := action["customerRef"].(string)
			meterName, _ := action["meterName"].(string)
			customer, err := customerSnapshotFromAction(action, backendRef)
			if err != nil {
				return nil, err
			}
			return &Allow{
				client:      c,
				backendRef:  backendRef,
				product:     opts.Product,
				meterName:   meterName,
				limits:      asObject(action["limits"]),
				customer:    customer,
				driverState: state,
			}, nil
		case "gate":
			if err := c.applyGateCache(action["cache"]); err != nil {
				return nil, err
			}
			if request, ok := action["request"].(map[string]any); ok && request != nil {
				if err := c.postUsage(ctx, request); err != nil {
					return nil, err
				}
			}
			gate, err := json.Marshal(action["gate"])
			if err != nil {
				return nil, fmt.Errorf("solvapay: gate_next gate: %w", err)
			}
			return &Paywall{Gate: gate}, nil
		default:
			return nil, &Error{Code: "internal_error", Message: "gate_next returned unknown action kind"}
		}
	}
}

func (c *Client) applyGateCache(raw any) error {
	cache, ok := raw.(map[string]any)
	if !ok || cache == nil {
		return nil
	}
	op, _ := cache["op"].(string)
	key, _ := cache["key"].(string)
	if key == "" {
		return nil
	}
	c.gate.mu.Lock()
	defer c.gate.mu.Unlock()
	switch op {
	case "delete":
		delete(c.gate.limitsCache, key)
	case "updateRemaining":
		entry := c.gate.limitsCache[key]
		entry.remaining = cache["remaining"]
		c.gate.limitsCache[key] = entry
	case "set":
		ts := time.UnixMilli(int64(asFloat(cache["timestamp"])))
		c.gate.limitsCache[key] = limitsCacheEntry{
			storedAt:  ts,
			remaining: cache["remaining"],
			limits:    asObject(cache["limits"]),
		}
	}
	return nil
}

func callDecisionJSON(ctx context.Context, fn string, args map[string]any) (json.RawMessage, error) {
	value, err := nativecall.CallValueJSON(ctx, fn, mustJSON(args))
	if err != nil {
		return nil, guestToError(err)
	}
	return value, nil
}

func guestToError(err error) error {
	var g *nativecall.GuestError
	if errors.As(err, &g) {
		return &Error{Code: g.Code, Message: g.Message, Status: g.Status, Retryable: g.Retryable}
	}
	return err
}

func mustJSON(v any) string {
	b, err := json.Marshal(v)
	if err != nil {
		panic(err)
	}
	return string(b)
}

func asObject(v any) map[string]any {
	if m, ok := v.(map[string]any); ok {
		return m
	}
	return map[string]any{}
}

func asBool(v any) bool {
	switch t := v.(type) {
	case bool:
		return t
	default:
		return false
	}
}

func asFloat(v any) float64 {
	switch t := v.(type) {
	case float64:
		return t
	case float32:
		return float64(t)
	case int:
		return float64(t)
	case int64:
		return float64(t)
	case json.Number:
		f, _ := t.Float64()
		return f
	default:
		return 0
	}
}
